### ISSUE TITLE
The reloader fails on SyntaxError when using a custom cli entry point

### ISSUE THREAD

**Issue by segevfiner** [pallets/flask#2424]
### Expected Behavior

The reloader should keep working even on SyntaxError when using a custom cli entry point. Obviously it might be diffcult to make it work as is, so it might require the cli entry point to be written in a certain way for this to work. For example it should be in a separate file that is not modified often during app development and should import the app only from the `create_app` function. This might just require documenting this more clearly.

```python
import click
from flask import Flask
from flask.cli import FlaskGroup

app = Flask(__name__)

@app.route('/')
def hello_world():
    return 'Hello, World!'

@click.group(cls=FlaskGroup, create_app=lambda info: app)
def cli():
    pass

if __name__ == '__main__':
    cli()
```

```cmd
set FLASK_DEBUG=1
python app.py run
```

### Actual Behavior

Introucing a SyntaxError into the file will cause the reloader to exit.

### Environment

* Python version: 2.7.13
* Flask version: 0.12.2
* Werkzeug version: 0.12.2

**Comment by untitaker** [comment]
This is technically not possible (briefly mentioned at the bottom of http://flask.pocoo.org/docs/0.12/server/) and one of the reasons the `flask` command exists.

**Comment by ThiefMaster** [comment]
If you do use a custom CLI command you might want to either keep the import path towards your cli entrypoint free from anything that is likely to break or simply making your CLI entry point a separate top-level module (`yourpackage_cli` for example). Then you define your CLI commands in that module and import the actual implementation from within the command function. That way the dev server won't die if you introduce syntax errors anywhere in your package (except something that causes the cli itself to fail at import time)

**Comment by segevfiner** [comment]
That is why I mentioned "documentation". I think we ought to at least leave a note about this in the documentation since we do have a section mentioning custom entry points and this limitation is not mentioned there.

**Comment by davidism** [comment]
That's not a custom entry point though, just a custom CLI. If you added the cli as an entry point in setup.py, you could use `myapp run` and get the correct behavior.

**Comment by untitaker** [comment]
Reopening as documentation issue.

**Comment by greyli** [comment]
I actually encounter a custom entry point problem, so I just post it here. 
I install the app with this setup:
```python
...
entry_points={
        'console_scripts': [
            'myapp=flask.cli:main',
        ],
    },
...
```

## Exception
When I run `myapp run`, the error below was displayed:
```pytb
(venv) C:\...>myapp run
Serving Flask app "devapp"
Forcing debug mode on
Restarting with stat
File "C:\...\venv\Scripts\myapp.exe", l
ine 1
SyntaxError: Non-ASCII character '\x90' in file C:\...\venv\Scripts\myapp.exe on line 1, but no encoding declared; see ht
tp://python.org/dev/peps/pep-0263/ for details
```
Everything is fine when I use `flask run` or `myapp shell`.

## Enviroment
- OS: windows 7
- Python version: 2.7.13
- Flask version: 0.12.2
- Werkzeug version: 0.12.2

Any ideas will be appreciated.

**Comment by ThiefMaster** [comment]
Unrelated, please open a new issue. In your case it tries to run a .exe file as python.

**Comment by greyli** [comment]
OK, thanks.

**Comment by segevfiner** [comment]
@greyli There is already a separate issue for that https://github.com/pallets/werkzeug/issues/1136, and a pull request https://github.com/pallets/werkzeug/pull/1138.

**Comment by greyli** [comment]
@segevfiner Got it, thanks.

**Comment by alanhamlett** [comment]
Maybe I should open a new issue, but want to check here first. I'm using the `flask run` command from latest docs, but if there's a `SyntaxError` then the dev server stops reloading.

This means even after the syntax error is fixed, the dev server doesn't reload and must be manually restarted.

**Comment by davidism** [comment]
@segevfiner would you re-test this with the latest version of Werkzeug? 0.14 should fix this.

**Comment by davidism** [comment]
@alanhamlett I'm fairly sure that's fixed with the latest Flask and Werkzeug masters.

**Comment by davidism** [comment]
@alanhamlett just tried it out, looks like it's fixed with Flask master and Werkzeug 0.14. That's when using `flask run`, not this issue.
